api/apiclient.go

```diff
@@ -637,17 +636,71 @@ func dialWebsocketMulti(ctx context.Context, addrs []string, path string, opts d
 	// Dial all addresses at reasonable intervals.
 	try := parallel.NewTry(0, nil)
 	defer try.Kill()
-	for _, addr := range addrs {
-		err := startDialWebsocket(ctx, try, addr, path, opts)
-		if err == parallel.ErrStopped {
+	// Make a context that's cancelled when the try
+	// completes so that (for example) a slow DNS
+	// query will be cancelled if a previous try succeeds.
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		<-try.Dead()
+		cancel()
+	}()
+	tried := make(map[string]bool)
+	var cacheUsed []string
+	for {
+		if len(addrs) == 0 && len(cacheUsed) > 0 {
+			// We've tried all the addresses but for some
+			// of them we used cached values which might
+			// have become out of date, so retry them
+			// with no cache.
+			addrs = cacheUsed
+			cacheUsed = nil
+			opts.DNSCache = emptyDNSCache{opts.DNSCache}
+		}
+		if len(addrs) == 0 {
 			break
 		}
+		addr := addrs[0]
+		addrs = addrs[1:]
+		host, port, err := net.SplitHostPort(addr)
 		if err != nil {
-			return nil, errors.Trace(err)
+			// Defensive - this should never happen because
+			// the addresses are checked with Info.Validate
+			// beforehand.
+			err := errors.Errorf("invalid address %q: %v", addr, err)
+			recordTryError(try, err)
+			continue
 		}
-		select {
-		case <-opts.Clock.After(opts.DialAddressInterval):
-		case <-try.Dead():
+		ips := opts.DNSCache.Lookup(host)
+		if len(ips) > 0 {
+			cacheUsed = append(cacheUsed, addr)
+		} else {
+			var err error
+			ips, err = lookupIPAddr(ctx, host, opts.IPAddrResolver)
+			if err != nil {
+				err := errors.Errorf("cannot resolve %q: %v", host, err)
+				recordTryError(try, err)
+				continue
+			}
+			opts.DNSCache.Add(host, ips)
+			logger.Debugf("looked up %v -> %v", host, ips)
```

Comment:
but that would mean that *this* would end up caching IP=>[IP].
It *definitely* doesn't make sense to put raw IP addrs into cacheUsed.